untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,493 @@
+
+### Church pairs
+
+Since we have booleans now, we can also encode pairs (and by extension, tuples as trees or right combs) as functions, returning one value when given $tru$ as an argument, and another when given $fls$.
+
+First of all, let us define a pair constructor combinator:
+
+$pair = \lambda f\;s\;b. b\;f\;s.$
+
+It looks suspiciously similar to our $ifThenElse$ combinator, the only difference being that condition is now the last argument.
+
+Now we can define two combinators for deconstructing a pair:
+
+$fst = \lambda p. p\;tru,$\
+$snd = \lambda p. p\;fls.$
+
+It's reasonably straightforward to show that these definitions work as intended.
+Consider, for instance,
+
+$fst\;(pair\;x\;y)$
+$= (\lambda p. p\;tru)\;((\lambda f\;s\;b. b\;f\;s)\;x\;y)$
+$\Rightarrow (\lambda p. p\;tru)\;(\lambda b. b\;x\;y)$
+$\to (\lambda b. b\;x\;y)\;tru$
+$\to tru\;x\;y$
+$= (\lambda a\; b. a)\;x\;y$
+$\Rightarrow x.$
+
+Here, double arrow $\Rightarrow$ signifies more than one step of beta-reduction.
+
+### Church naturals
+
+Since we encoded both pairs and booleans, we already can encode naturals simply as sequences of booleans, that is, as binary numbers.
+In practice, this turns out to be cumbersome, however, so instead, we'll define naturals similar to the Peano construction.
+Essentially, we'll be choosing an encoding for zero and then encoding other numbers as some successor function applied to zero.
+
+Church used the following definitions:
+
+$c_0 = \lambda s\;z. z\;$\
+$c_1 = \lambda s\;z. s\;z,$\
+$c_2 = \lambda s\;z. s\;(s\;z),$\
+etc.
+
+The beautiful thing about this encoding is that it naturally encodes the concept of "counting".
+Each Church natural takes a "next" combinator $s$ and a "starting value" $z,$ and then "counts" the corresponding "nexts" from $z.$
+
+Now we can define a successor function, taking a natural and returning the next natural.
+Since we defined naturals as functions of two arguments, we expect the successor function to be a function of three arguments (i.e. three nested abstractions).
+The body should apply the "next" argument to the natural one more time:
+
+$succ = \lambda n\; s\; z. s\;(n\;s\;z).$
+
+It might be easier to see what's going on if we explicitly note the outer abstraction:
+
+$succ = \lambda n. \lambda s\; z. s\;(n\;s\;z).$
+
+In the same manner, we can define addition
+$$plus = \lambda m\; n. \lambda s\; z. m\;s\;(n\;s\;z),$$
+and multiplication
+$$mul = \lambda m\; n. \lambda s\; z. m\;(n\;s)\;z.$$
+The former counts $m$ starting from $n$, and the latter counts $m$ times $n$ starting from $z.$
+
+We can also easily test whether a value corresponds to zero (i.e. $c_0$) or not:
+$$isZero = \lambda n. n\;(\lambda x. fls)\;tru.$$
+
+We run into a bit of an issue when trying to define subtraction or a predecessor function.
+Turns out it's doable but neither pretty nor efficient.
+We'll use pairs of values containing two subsequent naturals.
+That way, we can stagger the counter by one.
```

Comment:
Is that specific terminology? I don't really understand the use of this word here.

Reply:
Okay, I may be misusing the word here, but I definitely encountered "staggered counter" in the literature (meaning a counter lagging by some specific offset). Rewrote to

> That way, we can have a counter that's lagging by one.

Reply:
Oh, ok. Yes, that's definitely clearer.

Reply:
You used the word later as well, btw, that should also probably be changed. (On line 361 for me)